server/app/services/apikey/ApiKeyService.java

```diff
@@ -0,0 +1,278 @@
+      return form.withError("expiration", "Expiration cannot be blank.");
+    }
+
+    try {
+      Instant expiration =
+          DateConverter.parseIso8601DateToStartOfDateInstant(maybeExpirationString.get());
+      apiKey.setExpiration(expiration);
+    } catch (DateTimeParseException e) {
+      return form.withError("expiration", "Expiration must be in the form YYYY-MM-DD.");
+    }
+
+    return form;
+  }
+
+  private DynamicForm resolveSubnet(DynamicForm form, ApiKey apiKey) {
+    Optional<String> maybeSubnetString =
+        Optional.ofNullable(form.rawData().getOrDefault("subnet", null));
+
+    if (!maybeSubnetString.isPresent() || maybeSubnetString.get().isBlank()) {
+      return form.withError("subnet", "Subnet cannot be blank.");
+    }
+
+    try {
+      new SubnetUtils(maybeSubnetString.get());
+    } catch (IllegalArgumentException e) {
+      return form.withError("subnet", "Subnet must be in CIDR notation.");
+    }
+
+    apiKey.setSubnet(maybeSubnetString.get());
+
+    return form;
+  }
+
+  private static final Pattern GRANT_PROGRAM_READ_PATTERN =
+      Pattern.compile("^grant-program-read\\[([\\w\\-]+)\\]$");
+
+  private DynamicForm resolveGrants(DynamicForm form, ApiKey apiKey)
+      throws ProgramNotFoundException {
+    ImmutableSet<String> programSlugs = programService.getAllProgramSlugs();
+
+    for (String formDataKey : form.rawData().keySet()) {
+      Matcher matcher = GRANT_PROGRAM_READ_PATTERN.matcher(formDataKey);
+
+      if (!matcher.find()) {
+        continue;
+      }
+
+      matcher.matches();
+
+      String programSlug = matcher.group(1);
+
+      if (!programSlugs.contains(programSlug)) {
+        throw new ProgramNotFoundException(programSlug);
+      }
+
+      apiKey.getGrants().grantProgramPermission(programSlug, Permission.READ);
+    }
+
+    return form;
+  }
+
+  private String generateSecret(int length) {
+    KeyGenerator keyGen;
+
+    try {
+      keyGen = KeyGenerator.getInstance("AES");
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
+
+    keyGen.init(length);
+
+    byte[] secret = keyGen.generateKey().getEncoded();
+    return Base64.getEncoder().encodeToString(secret);
+  }
+
+  private String salt(String message) {
+    byte[] rawMessage = Base64.getDecoder().decode(message);
+    byte[] rawKey = Base64.getDecoder().decode(secretSalt);
+
+    HashFunction hashFunction = Hashing.hmacSha256(rawKey);
+    HashCode saltedMessage = hashFunction.hashBytes(rawMessage);
+
+    return Base64.getEncoder().encodeToString(saltedMessage.asBytes());
+  }
+
+  private String getAuthorityId(CiviFormProfile profile) {
+    String authorityId = profile.getAuthorityId().join();
+
+    if (authorityId != null) {
+      return authorityId;
+    }
+
+    if (environment.isDev()) {
+      return "dev-admin";
+    }
+
+    throw new RuntimeException("ApiKey creator must have authority_id.");
```

Comment:
How does play handle Runtime Exceptions?  Is this something that should returned a permission_denied or similar error code?

Reply:
Unhandled exceptions result in a 500 status code.

In the circumstance here, there should always be an authority_id. Only admins can call this method, and all admins should have an authority_id in prod. In dev and test we can use a dummy value. If this code somehow runs in prod and is passed a profile that doesn't have an authority_id, that's a violation of an invariant in the system and a server error we should look into.